tools/walletextension/api/routes.go

```diff
@@ -216,3 +241,124 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
+		return
+	}
+
+	accAddress := gethcommon.HexToAddress(reqJSONMap[common.JSONKeyAddress])
+
+	signature, err := hex.DecodeString(reqJSONMap[common.JSONKeySignature][2:])
+	if err != nil {
+		userConn.HandleError(fmt.Sprintf("could not decode signature from client to hex: %s", err))
+		return
+	}
+
+	message, ok := reqJSONMap[common.JSONKeyMessage]
+	if !ok || message == "" {
+		userConn.HandleError("message not found in the request")
+		return
+	}
+
+	userID, err := getUser(userConn.ReadRequestParams())
+	if err != nil {
+		userConn.HandleError("userID not found in the request")
+		return
+	}
+
+	// check the userID corresponds to the one in text
+	messageUserID := ""
+	// todo @ziga ( check if messageAddress needs to be included in the request)
+	// messageAddress := ""
+	regex := regexp.MustCompile(`^Register\s(\w+)\sfor\s(\w+)$`)
+	if regex.MatchString(message) {
+		params := regex.FindStringSubmatch(message)
+		messageUserID = params[1]
+		// messageAddress = params[2]
+	} else {
+		userConn.HandleError(fmt.Sprintf("Submitted message is not in the correct format: %s", message))
+	}
+
+	if userID != messageUserID || messageUserID == "" {
+		userConn.HandleError(fmt.Sprintf("User in submitted message (%s) does not match user provided in the request (%s)", messageUserID, userID))
+	}
+
+	// check the signature if it corresponds to the address and is valid
+	vk, found := walletExt.UnsignedVKs[accAddress]
+	if !found {
+		userConn.HandleError(fmt.Sprintf("no viewing key found to sign for acc=%s, please visit /join/ before sending signature", accAddress))
+		return
+	}
+
+	// We transform the V from 27/28 to 0/1. This same change is made in Geth internals, for legacy reasons to be able
+	// to recover the address: https://github.com/ethereum/go-ethereum/blob/55599ee95d4151a2502465e0afc7c47bd1acba77/internal/ethapi/api.go#L452-L459
+	signature[64] -= 27
+
+	messageHash := crypto.Keccak256Hash([]byte(message))
+	signatureNoRecoverID := signature[:len(signature)-1]
+	verified := crypto.VerifySignature(vk.PublicKey, messageHash.Bytes(), signatureNoRecoverID)
+
+	if !verified {
+		userConn.HandleError("signature verification was not successful for your account")
+		return
+	}
+
+	// save the text+signature against the userID
+	// todo: where should I save the text? In another column in the database?
+	vk.SignedKey = signature
+	err = walletExt.Storage.SaveUserVK(userID, vk, message)
+	if err != nil {
+		userConn.HandleError("error saving viewing key")
+		return
+	}
+}
+
+func joinRequestHandler(walletExt *walletextension.WalletExtension, userConn userconn.UserConn) {
+	// todo (@ziga) add protection against DDOS attacks
+	_, err := userConn.ReadRequest()
+	if err != nil {
+		return
+	}
+
+	// generate new key-pair
+	viewingKeyPrivate, err := crypto.GenerateKey()
+	viewingPrivateKeyEcies := ecies.ImportECDSA(viewingKeyPrivate)
+	if err != nil {
+		userConn.HandleError(fmt.Sprintf("could not generate new keypair: %s", err))
```

Comment:
these errors should be returned as "Internal error" to the user, and logged with this detailed message.